fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -4620,6 +4620,9 @@ public static class Builder implements BaseBuilder<Message, FDBRecordStore> {
         @Nullable
         private FDBRecordStoreStateCache storeStateCache = null;
 
+        @Nullable
+        private Boolean versionChanged;
```

Comment:
My idea was to make the version change information available in case the consumer is calling vanilla versions of `create()`, `open()`, `createOrOpen` - which all route to `createOrOpenAsync()` and also "guarantees" checking version on startup. 

I am not quite sure of the contract here though, but this is what I perceive. `build()` is essentially doing the bare minimum thing of unintelligently initializing the store and `unchecked__()` are indicative of skipping version checks IMO. If the consumer goes on to call `checkVersion` on the returned store, the version changing information is readily available. However, extending your point, there is definitely info loss if there are multiple calls to `build()` / `createOrOpen()`, which is unusual in general. 

Maybe we can just make the API to be such that the versionChanged information is available _only_ when the store is initialized with checks and is not guaranteed to be always in sync. 

I am fine doing it the other way but the contract there will be more unclear as checkVersion() is heavily called in conjunction with other basic operations of the store and version change information can be lost even there.

Reply:
Yeah, that's right when it comes to the contract of `build()` and `uncheckedOpen()`. They are mainly there so that users who want to have more control over when check version gets called, but it's debatable how wise they are, given how important it is to call `checkVersion`.

The way I was imagining it, `versionChanged` on the store would be updated at the end of `checkVersion`, and it would be guaranteed to be in sync if and only if (1) a `checkVersion` call has completed and (2) no other `checkVersion` call has started. If we really cared about cases where we call `checkVersion` multiple times, we could also make it a counter instead of a Boolean, though that seems like it could be overkill.

What kind of loss-of-version-change information are you envisioning? Just the case where check version is called multiple times, or something else? To my knowledge, the logic in `FDBRecordStore` that makes sure `checkVersion` has been called (such as it is) won't try and call it a second time if it has already been called.

On the other hand, my reservation about putting this information in the builder is mainly that people don't expect builders to retain state from the thing they've built, and in particular, use cases that re-use builders (perhaps swapping out the transaction to open otherwise identical record stores with different transactions) can unexpectedly carry over state between calls to `open()` (and variants).

Reply:
That's exactly what I am worried about, the consumer calls `checkVersion` multiple times. In that case, we might lose the relevant state of the flag from a previous `checkVersion` call. I think this doesn't happen in the `FDBRecordStore` as such, but we should be wary of it being misused. 

However, we can decide on the contract such as`VersionChanged` is true if there is a version update at least once in the lifetime of the transaction (and `FDBRecordStore`).